IPython/core/completer.py

```diff
@@ -2360,11 +2360,122 @@ def _jedi_matches(
             else:
                 return iter([])
 
+    class _CompletionContextType(enum.Enum):
+        ATTRIBUTE = "attribute"  # For attribute completion
+        GLOBAL = "global"  # For global completion
+
+    def _determine_completion_context(self, line):
+        """
+        Determine whether the cursor is in an attribute or global completion context.
+        """
+        # Cursor in string/comment → GLOBAL.
+        if self._is_in_string_or_comment(line):
+            return self._CompletionContextType.GLOBAL
+
+        # Match 'anything-dot-word' at end for attribute context.
+        # Ex: 'obj.', 'np.random.ran'.
+        chain_match = re.search(r"([a-zA-Z_].*)\.([a-zA-Z_][a-zA-Z0-9_]*)?$", line)
+        if chain_match:
+            return self._CompletionContextType.ATTRIBUTE
+
+        # No attribute pattern → GLOBAL.
+        # Ex: 'var', '', '3.', '-42.5.'
+        return self._CompletionContextType.GLOBAL
```

Comment:
Some edge cases not covered, probably not important enough to cover, but listing them for completness:

```
(3).to_<tab>
```

expecting `(3).to_bytes`

Note that `3.to_bytes` is invalid in Python

```
3.1.as_<tab>
-3.1.r<tab>
```

Expecting `3.1.as_integer_ratio` and `-3.1.real` respectively.